crates/dids/src/method/dht/document_packet/service.rs

```diff
@@ -0,0 +1,222 @@
+use std::collections::HashMap;
+
+use crate::document::Service;
+
+use simple_dns::{
+    rdata::{RData, TXT},
+    Name, ResourceRecord,
+};
+
+use super::{DocumentPacketError, DEFAULT_TTL};
+
+/// Gets the RData from the record. If RData is RData::TXT, get the text as a string.
+/// Convert strings like "id=foo;t=bar;se=baz" into a map like { 'id': 'foo', 't': 'bar', 'se': 'baz' }
+/// If there is any issue, return DocumentPacketError::RDataError
+fn record_rdata_to_hash_map(
+    record: ResourceRecord,
+) -> Result<HashMap<String, String>, DocumentPacketError> {
+    // Get RData text as String
+    let rdata_txt = match record.rdata {
+        RData::TXT(txt) => txt,
+        _ => {
+            return Err(DocumentPacketError::RDataError(
+                "RData must have type TXT".to_owned(),
+            ))
+        }
+    };
+    let text = match String::try_from(rdata_txt) {
+        Ok(text) => text,
+        Err(_) => {
+            return Err(DocumentPacketError::RDataError(
+                "Failed to convert to string".to_owned(),
+            ))
+        }
+    };
+
+    // Parse key-value pairs:
+    //   Split string by ";" to get entries
+    //   Split each entry by "=" to get key and value
+    let mut attributes = HashMap::new();
+    for entry in text.split(';') {
+        let k_v: Vec<&str> = entry.split('=').collect();
+        if k_v.len() != 2 {
+            return Err(DocumentPacketError::RDataError(
+                "Could not get values from RData text".to_owned(),
+            ));
+        }
+
+        let k = k_v[0].trim().to_string();
+        let v = k_v[1].trim().to_string();
+
+        attributes.insert(k, v);
+    }
+
+    Ok(attributes)
+}
+
+/// Get value from the RData HashMap created by record_rdata_to_hash_map().
+/// Convert `None` into DocumentPacketError
+fn get_rdata_txt_value(
+    rdata_map: &HashMap<String, String>,
+    key: &str,
+) -> Result<String, DocumentPacketError> {
+    let val = rdata_map
+        .get(key)
+        .ok_or(DocumentPacketError::RDataError(format!(
+            "Could not extract {} from RData",
+            key
+        )))?;
+    Ok(val.to_string())
+}
+
+impl Service {
+    pub fn to_resource_record(
+        &self,
+        idx: u32,
+    ) -> Result<ResourceRecord<'static>, DocumentPacketError> {
+        let service_id_fragment = self
```

Comment:
do we want to throw here if split length is more than 2?

Reply:
I decided instead we should return everything after the first `#` instead of returning everything after the final `#`. The URL spec seems to say that addition `#` in a fragment are invalid, but upon running the url crate and node `url` it looks like additional `#` are included as part of the fragment. I don't want to reinvent the wheel, so I'm gonna use the `url` crate and take its default behavior sufficient.